### ISSUE TITLE
vllm-omni integration for 2.8x speedup

### ISSUE THREAD

**Issue by lzqlzzq** [multimodal-art-projection/YuE#175]
Hi M.A.P team,

Thanks for the model! I tried integrated the model into [vllm-omni](https://github.com/vllm-project/vllm-omni) at [lzqlzzq/vllm-omni](https://github.com/lzqlzzq/vllm-omni). It yields approximately 2.8x speedup!

Speed benchmark:
| Prompt / mode / seed | Backend | ABC / semantic content tokens | Audio (s) | Request (s) | RTF |
|---|---|---:|---:|---:|---:|
| English acoustic folk / full / 831001 | Original | 843 / 1976 | 79.04 | 91.74 | 1.161 |
| Same input | Omni | 761 / 1746 | 69.84 | 28.33 | 0.406 |
| English indie rock / melody / 831002 | Original | 610 / 1832 | 73.28 | 76.73 | 1.047 |
| Same input | Omni | 610 / 1820 | 72.80 | 27.91 | 0.383 |
| Mandarin folk / full / 831005 | Original | 519 / 1509 | 60.36 | 63.72 | 1.056 |
| Same input | Omni | 519 / 1870 | 74.80 | 27.78 | 0.371 |
| Mandarin R&B / melody / 831006 | Original | 342 / 1082 | 43.28 | 45.71 | 1.056 |
| Same input | Omni | 522 / 1574 | 62.96 | 24.11 | 0.383 |

| Stage | Original | Omni |
|---|---:|---:|
| ABC / semantic content tokens | 2314 / 6399 | 2412 / 7010 |
| Generated audio | 255.95 s | 280.39 s |
| Sum of request service times | 277.91 s | 108.14 s |
| Weighted RTF: sum(time) / sum(audio) | 1.086 | 0.386 |
| Audio seconds / request wall second | 0.921 | 2.593 |

VRAM costs `I am working on it :(   `:

| Backend | Observed peak MiB | Observed peak GiB | Samples |
|---|---:|---:|---:|
| Original | 8882 | 8.673828 | 1944 |
| Omni | 20029 | 19.559570 | 1944 |

AR logits parity (seems to have some difference, but in my listening test it just well):

| metrics | value |
|---|---:|
| ABC top1 | 52/52 |
| ABC KL | 0.00128309713 |
| ABC KL | 0.03110020934 |
| semantic top1 | 17/17 |
| mean semantic KL | 0.00024639285 |
| max semantic KL | 0.00091689777 |

VAE parity:
| Mode | SNR (dB) | SI-SDR (dB) | Residual RMS (dBFS) |
|---|---:|---:|---:|
| full / external ABC | 120.10 | 120.10 | −141.93 |
| full / automatic ABC | 122.20 | 122.20 | −139.99 |
| melody / external ABC | 122.92 | 122.92 | −146.15 |
| melody / automatic ABC | 123.40 | 123.40 | −145.91 |

Moreover, I have some questions:
1. Would it make sense to split the AR and NAR paths into two explicit modules for inference? The AR module could output the semantic tokens plus per-layer KV, and the NAR flow module just consume those cached KVs directly. There is no actually cross interaction between AR stream and NAR stream.
2. Would you provide the exact evaluation pipeline for me to do evaluation in the optimized implementation? If that's inconvient, you may use the fork to evaluate yourself for me to see the parity.
3. Will there further breaking change of the model architecture?

**Comment by lzqlzzq** [comment]
## Update(2026.9.14)

Environment:
- Hardwre: single RTX 3090 24 GiB
- Software: PyTorch 2.13.0, CUDA 13 toolchain, vLLM/Omni 0.28.0, Transformers 5.14.1, tiktoken 0.14.0.
- Config: TF32 disabled with NVIDIA_TF32_OVERRIDE=0.

| Aggregate metric, four measured songs | Original YuE2 | Final vLLM-Omni |
|---|---:|---:|
| Total request time | 276.19 s | 107.75 s |
| Generated audio | 255.95 s | 280.39 s |
| Aggregate RTF, lower is better | 1.079 | 0.384 |
| Audio seconds per wall-clock second | 0.93 | 2.60 |
| Serial song throughput | 0.87/min | 2.23/min |
| Effective ABC TPS | 38.55 | 137.50 |
| Effective semantic TPS | 38.52 | 137.56 |
| Maximum sampled request device memory | 8.673 GiB | 16.965 GiB |

Batching also implemented, it would yield 1-1.5x throughput depend on batch size.
